File: src/lunar_birthday_ical/lunar.py
import datetime
import logging

from lunar_python import Lunar, LunarYear

logger = logging.getLogger(__name__)


def get_future_lunar_equivalent_date(
    solar_datetime: datetime.datetime, age: int
) -> datetime.datetime:
    """
    Calculate the equivalent future solar date for a given past solar date and a target lunar year.
    """
    # 计算给定 公历日期 对应的 农历日期
    # .fromDate 所接受的类型为 datetime.datetime, 实际上处理后会把 time 部分丢弃
    lunar = Lunar.fromDate(solar_datetime)
    year = lunar.getYear() + age
    lunar_year = LunarYear.fromYear(year)

    # 获取闰月
    leap_month = lunar_year.getLeapMonth()

    # 确定目标年份的农历月
    if lunar.getMonth() > 0:
        lunar_month = lunar_year.getMonth(lunar.getMonth())
    elif abs(lunar.getMonth()) == leap_month:
        lunar_month = lunar_year.getMonth(lunar.getMonth())
    else:
        lunar_month = lunar_year.getMonth(abs(lunar.getMonth()))

    # 确定农历日
    lunar_day = min(lunar.getDay(), lunar_month.getDayCount())

    # 创建目标年份的农历日期
    future_lunar = Lunar.fromYmd(year, lunar_month.getMonth(), lunar_day)

    # 转换为公历日期, 恢复原本的时间和 timezone
    solar_time = solar_datetime.time()
    future_solar_datetime = datetime.datetime.strptime(
        future_lunar.getSolar().toYmd(), "%Y-%m-%d"
    ).replace(
        hour=solar_time.hour,
        minute=solar_time.minute,
        second=solar_time.second,
        tzinfo=solar_datetime.tzinfo,
    )

    return future_solar_datetime

File: web/src/lib/lunar.ts
import { Lunar, LunarYear, Solar } from 'lunar-typescript';

export function getFutureSolarDate(solarDateStr: string, targetYear: number): Date {
  // solarDateStr is YYYY-MM-DD
  const [year, month, day] = solarDateStr.split('-').map(Number);
  const solar = Solar.fromYmd(year, month, day);
  const lunar = solar.getLunar();

  const targetLunarYear = LunarYear.fromYear(targetYear);
  const leapMonth = targetLunarYear.getLeapMonth();

  let targetLunarMonth;
  const lunarMonth = lunar.getMonth();

  // Logic from python:
  // if lunar_date.getMonth() > 0: target_lunar_month = target_lunar_year.getMonth(lunar_date.getMonth())
  // elif abs(lunar_date.getMonth()) == leap_month:
  //     target_lunar_month = target_lunar_year.getMonth(lunar_date.getMonth())
  // else: target_lunar_month = target_lunar_year.getMonth(abs(lunar_date.getMonth()))

  // In lunar-typescript, getMonth() returns negative for leap month?
  // Let's check standard behavior. Usually negative means leap.

  if (lunarMonth > 0) {
    targetLunarMonth = targetLunarYear.getMonth(lunarMonth);
  } else if (Math.abs(lunarMonth) === leapMonth) {
    targetLunarMonth = targetLunarYear.getMonth(lunarMonth);
  } else {
    targetLunarMonth = targetLunarYear.getMonth(Math.abs(lunarMonth));
  }

  if (!targetLunarMonth) {
    // Fallback if something goes wrong, though shouldn't happen with valid inputs
    throw new Error(`Could not find target lunar month for year ${targetYear}`);
  }

  const targetLunarDay = Math.min(lunar.getDay(), targetLunarMonth.getDayCount());

  // Note: Lunar.fromYmd takes (lunarYear, lunarMonth, lunarDay)
  // If month is negative, it handles it as leap month if valid.
  const targetLunarDate = Lunar.fromYmd(targetYear, targetLunarMonth.getMonth(), targetLunarDay);

  const targetSolar = targetLunarDate.getSolar();
  return new Date(targetSolar.getYear(), targetSolar.getMonth() - 1, targetSolar.getDay());
}

export function solarToLunar(dateStr: string): string {
  const [year, month, day] = dateStr.split('-').map(Number);
  const solar = Solar.fromYmd(year, month, day);
  const lunar = solar.getLunar();
  return lunar.toString();
}

export function lunarToSolar(year: number, month: number, day: number, isLeap: boolean): string {
  // This is a bit tricky without a direct "Lunar to Solar" input form that matches exactly.
  // But usually we start from Solar in this app.
  // If we need a converter, we can implement it.
  const lunar = Lunar.fromYmd(year, isLeap ? -month : month, day);
  const solar = lunar.getSolar();
  return solar.toYmd();
}
